### PR TITLE
add redirect support and request retry

### PR DESCRIPTION
**Pull request by tuxlife**
please review

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[report]
+# Regexes for lines to exclude from consideration
+exclude_lines =
+    # Have to re-enable the standard pragma
+    pragma: no cover
+
+    # Don't complain about missing debug-only code:
+    def __repr__
```
Comment (hjacobs) [inline .coveragerc]: why do we need to exclude more lines here?

**Comment by tuxlife** [comment]
Fix for #13

File: berry/cli.py [hunk 2]
```diff
@@ -59,58 +59,86 @@ def run_berry(args):
...
+                            region = 'eu-central-1'
+                            logging.error(('Invalid Request while trying to read "{}" from mint S3 bucket "{}". ' +
+                                           'Retry with region {}! ' +
+                                           '(S3 error message: {})').format(
+                                          key_name, mint_bucket, region, msg))
+                            s3 = session.client('s3', 'eu-central-1')
+                        elif error_code == 'PermanentRedirect' and endpoint.endswith('.amazonaws.com'):
+                            endpoint_parts = endpoint.split('.')
+                            region = endpoint_parts[-3].replace('s3-', '')
+                            logging.error(('Get Redirect while trying to read "{}" from mint S3 bucket "{}". ' +
```
Comment (hjacobs) [inline berry/cli.py]: "Got redirect" (past tense)

File: berry/cli.py [hunk 3]
```diff
@@ -59,58 +59,86 @@ def run_berry(args):
...
+                        status_code = e.response.get('ResponseMetadata', {}).get('HTTPStatusCode')
+                        msg = e.response['Error'].get('Message')
+                        error_code = e.response['Error'].get('Code')
+                        endpoint = e.response['Error'].get('Endpoint', '')
+                        retry -= 1
+                        if error_code == 'InvalidRequest':
+                            # Try with a SigV4 Region
+                            region = 'eu-central-1'
+                            logging.error(('Invalid Request while trying to read "{}" from mint S3 bucket "{}". ' +
+                                           'Retry with region {}! ' +
```
Comment (hjacobs) [inline berry/cli.py]: "Retrying with region .."

File: berry/cli.py [hunk 4]
```diff
@@ -59,58 +59,86 @@ def run_berry(args):
...
+                                           '(S3 error message: {})').format(
+                                          key_name, mint_bucket, region, msg))
+                            s3 = session.client('s3', 'eu-central-1')
+                        elif error_code == 'PermanentRedirect' and endpoint.endswith('.amazonaws.com'):
+                            endpoint_parts = endpoint.split('.')
+                            region = endpoint_parts[-3].replace('s3-', '')
+                            logging.error(('Get Redirect while trying to read "{}" from mint S3 bucket "{}". ' +
+                                           'Retry with region {}! ' +
```
Comment (hjacobs) [inline berry/cli.py]: "Retrying with region .."

File: berry/cli.py [hunk 5]
```diff
@@ -59,58 +59,86 @@ def run_berry(args):
...
+                        retry = False
+                    except botocore.exceptions.ClientError as e:
+                        # more friendly error messages
+                        # https://github.com/zalando-stups/berry/issues/2
+                        status_code = e.response.get('ResponseMetadata', {}).get('HTTPStatusCode')
+                        msg = e.response['Error'].get('Message')
+                        error_code = e.response['Error'].get('Code')
+                        endpoint = e.response['Error'].get('Endpoint', '')
+                        retry -= 1
+                        if error_code == 'InvalidRequest':
```
Comment (hjacobs) [inline berry/cli.py]: there are probably other errors (non-signature errors) with code "InvalidRequest"

File: berry/cli.py [hunk 6]
```diff
@@ -59,58 +59,86 @@ def run_berry(args):
...
+                            # Try with a SigV4 Region
+                            region = 'eu-central-1'
```
Comment (hjacobs) [inline berry/cli.py]: What about more regions with sigv4? Having the region hardcoded here sounds wrong

**Comment by hjacobs** [comment]
:+1:

**Pull request merged**